app/controllers/restaurants_controller.rb

```diff
@@ -1,16 +1,62 @@
+require_relative './application_controller'
+require 'active_record'
+
 class RestaurantsController < ApplicationController
-  def list
+  before_action :set_restaurant, only: [:show, :update, :destroy]
+  #TODO: consider adding authentication
+  rescue_from ActiveRecord::RecordNotFound, :with => :invalid_id_handler
+
+  # GET /restaurants
+  def index
+    @restaurants = Restaurant.all
+    @restaurants = @restaurants.where('cuisine = ?', query_params["by_cuisine"]) if query_params.has_key? "by_cuisine"
+    @restaurants = @restaurants.where('rating >= ?', query_params["min_rating"].to_i) if query_params.has_key? "min_rating"
+    @restaurants = @restaurants.where('delivery_time <= ?', query_params["max_delivery_time"].to_i) if query_params.has_key? "max_delivery_time"
+    render :json => @restaurants
   end
 
-  def retrieve
+  # GET /restaurants/:id
+  def show
+    @restaurant
+    render :json => @restaurant
   end
 
+  #POST /restaurants
   def create
+    @restaurant = Restaurant.create(restaurant_params)
```

Comment:
Consider doing the following:
Calling `create!` and then rescuing from an error. The line would be like so:
```
begin
    @restaurant = Restaurant.create!(restaurant_params)
      render status: :created, json: {
          message: 'The restaurant was added successfully :)'
      }.to_json
 rescue StandardError => e
      render status: :bad_request, json: {
          message: 'Failed to add a restaurant. Please try again :('
      }.to_json
end
```
create! (and basically all the DB function with ! at the end) don't fail silently, they throw an exception.
That way you don't need to check if it's `valid?`, since if it isn't, an error will be thrown by the model (something like: `ActiveRecord::RecordInvalid `)